wp-includes/class-wp-service-workers.php

```diff
@@ -83,6 +187,149 @@ public function register( $handle, $src, $deps = array(), $scope = self::SCOPE_A
+				'route'    => $route, // @todo route is Not the best name in case of array of routes for precaching.
+				'strategy' => $strategy,
+			);
+		} elseif ( ! is_string( $route ) ) {
+			/* translators: %s is caching strategy */
+			$error = sprintf( __( 'Route for the caching strategy %s must be a string.', 'pwa' ), $strategy );
+			_doing_it_wrong( __METHOD__, esc_html( $error, 'pwa' ), '0.2' );
+		} else {
+			$this->registered_caching_routes[] = array(
+				'route'    => $route,
+				'strategy' => $strategy,
+				'args'     => $args,
+			);
+		}
+	}
+
+	/**
+	 * Register precaching for routes. Gets the hashes of files' contents and adds as revision for each route.
+	 *
+	 * @param array $routes Array of routes.
+	 * @return string Precaching logic.
+	 */
+	protected function register_precaching_for_routes( $routes ) {
+
+		$routes_list = '[';
+
+		foreach ( $routes as $route ) {
+			$validated_path = $this->get_validated_file_path( $route, false );
+			if ( ! is_wp_error( $validated_path ) ) {
+				$file_content = @file_get_contents( $validated_path ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged, WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents, WordPress.WP.AlternativeFunctions.file_system_read_file_get_contents
+				if ( ! $file_content ) {
+					continue;
+				}
+
+				$hash = md5( $file_content );
+				if ( '[' !== $routes_list ) {
+					$routes_list .= ',';
+				}
+				$routes_list .= '
+{
+	url: ' . wp_json_encode( $route ) . ',
+	revision: ' . wp_json_encode( $hash ) . ',
+}';
+			}
+		}
+
+		if ( '[' === $routes_list ) {
+			return '';
+		}
+
+		$routes_list .= ']';
+
+		return 'wp.serviceWorker.precaching.precacheAndRoute(' . $routes_list . ");\n";
+	}
+
+	/**
+	 * Register caching strategy for route.
+	 *
+	 * @param string $route Route.
+	 * @param int    $strategy Caching strategy.
+	 * @param array  $args Array of args, can be cache_name, max_age, max_entries.
+	 * @return string Script.
+	 */
+	protected function register_caching_strategy_for_route( $route, $strategy, $args ) {
+		$script = '';
+
+		if ( isset( $args['cache_name'] ) ) {
+			$script .= 'const args = {
+	cacheName: ' . wp_json_encode( $args['cache_name'] ) . '
+};';
+		}
+		if ( isset( $args['max_entries'] ) || isset( $args['max_age'] ) ) {
+			$script .= 'args.plugins = [
+	new wp.serviceWorker.expiration.Plugin({';
+			if ( isset( $args['max_age'] ) ) {
+				$script .= '
+		maxAgeSeconds: ' . wp_json_encode( $args['max_age'] );
+				$script .= isset( $args['max_entries'] ) ? ',' : '';
+			}
+			if ( isset( $args['max_entries'] ) ) {
+				$script .= '
+		maxEntries: ' . wp_json_encode( $args['max_entries'] );
+			}
+			$script .= '
+	})
+];';
+		}
+
+		$args_script = $script;
+
+		$script .= '
+wp.serviceWorker.routing.registerRoute(';
+
+		// Check if the route is regex or string literal.
+		$is_regex_string = @preg_match( $route, '' ); // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged
```

Comment:
Is there a better way how to check if it's a `string` or regex?

Reply:
I think we need to explicitly indicate whether a given route is a regex. For example, in `$args` there could be `'regex' => true`.

Alternatively, we could automatically treat a `$route` that begins and ends with `/` as a regex, and then `trim('$route, /')` when passing into the `RegExp` constructor. But this isn't good either because, of course, URLs often begin and end with `/` 😄 

So I think explicitly indicating the route type via a flag is the way to go.

Reply:
Added within c16b6b5, defaulting to `false`.